# 🍳 Personal Chef AI Agent
### Module 1 Project — LangChain Foundations

An AI agent that suggests recipes based on your available ingredients.  
You can interact with it by **typing your ingredients** or **uploading a photo of your fridge**.

Built with: LangChain · LangGraph · Google Gemini · Tavily · BeautifulSoup

## PHASE 1: Setup & Configuration

In [1]:
from dotenv import load_dotenv
import os

path = r'C:\Users\user\lca-lc-foundations\.env'
load_dotenv(dotenv_path = path) 

True

In [2]:
from langchain_google_genai import ChatGoogleGenerativeAI

# Using Gemini for its vision capabilities (image understanding)
model = ChatGoogleGenerativeAI(model = "gemini-2.5-flash")

## Phase 2: Tools

The agent has two tools:
- **web_search** — finds recipes on the web based on ingredients
- **get_page_content** — fetches the full instructions from a recipe URL

Without `get_page_content`, the agent would only return short snippets and links instead of complete recipe steps.

In [4]:
from langchain.tools import tool 
from tavily import TavilyClient

tavily = TavilyClient()

@tool
def web_search(query: str) :
    """Search the web for cooking recepies"""
    return tavily.search(query) 
response = web_search.invoke("What are te best recepies in the world")  #should use .invoke with tools ! 

In [5]:
print (response)

{'query': 'What are te best recepies in the world', 'follow_up_questions': None, 'answer': None, 'images': [], 'results': [{'url': 'https://www.allrecipes.com/worlds-best-recipes-8787129', 'title': "22 'World's Best' Recipes That Our Community Can't Get Enough Of", 'content': 'Our list of 22 recipes includes everything from pork chops and soups to pastries, funnel cakes, and fudge, all claiming to be the best.', 'score': 0.6100127, 'raw_content': None}, {'url': 'https://www.tasteatlas.com/best-recipes-in-the-world', 'title': 'Top 100 Traditional Recipes in the World - TasteAtlas', 'content': 'Top 100 Traditional Recipes in the World · Vori-vori · Conchitas a la parmesana · Zhēngjiǎo · Karantika · Otoro nigiri sushi · Chutoro nigiri sushi.', 'score': 0.6002124, 'raw_content': None}, {'url': 'https://foreverlostintravel.com/best-recipes-from-around-the-world/', 'title': 'Best Recipes From Around the World - Forever Lost In Travel', 'content': 'That’s why I thought what better way than to

In [6]:
from pprint import pprint 

pprint (response)

{'answer': None,
 'follow_up_questions': None,
 'images': [],
 'query': 'What are te best recepies in the world',
 'request_id': 'd527b778-02c5-4738-ad07-8713ddd2d8a9',
 'response_time': 0.86,
 'results': [{'content': 'Our list of 22 recipes includes everything from pork '
                         'chops and soups to pastries, funnel cakes, and '
                         'fudge, all claiming to be the best.',
              'raw_content': None,
              'score': 0.6100127,
              'title': "22 'World's Best' Recipes That Our Community Can't Get "
                       'Enough Of',
              'url': 'https://www.allrecipes.com/worlds-best-recipes-8787129'},
             {'content': 'Top 100 Traditional Recipes in the World · Vori-vori '
                         '· Conchitas a la parmesana · Zhēngjiǎo · Karantika · '
                         'Otoro nigiri sushi · Chutoro nigiri sushi.',
              'raw_content': None,
              'score': 0.6002124,
              'titl

In [7]:
import json
print(json.dumps(response, indent=2, ensure_ascii=False))

{
  "query": "What are te best recepies in the world",
  "follow_up_questions": null,
  "answer": null,
  "images": [],
  "results": [
    {
      "url": "https://www.amazon.com/Best-Recipes-World-Mark-Bittman/dp/0767906721",
      "title": "The Best Recipes in the World: Bittman, Mark - Amazon.com",
      "content": "A comprehensive global cookbook featuring over 1,000 accessible recipes from dozens of countries, including European, Asian, African, Middle Eastern, and",
      "score": 0.99998605,
      "raw_content": null
    },
    {
      "url": "https://www.youtube.com/watch?v=IOik-6yXto0",
      "title": "We Ranked the BEST Recipes in the World and the Results Will ...",
      "content": "Inside the Kitchen is back with this week's top ten recipe list! These home cooks stepped up big time with mouthwatering recipes that just",
      "score": 0.99994075,
      "raw_content": null
    },
    {
      "url": "https://www.tasteatlas.com/best-recipes-in-the-world",
      "title": "Top 1

In [9]:
def show_full_response(response_dict):
    """Print the ENTIRE response with no truncation"""
    print(f"🔍 Query: {response_dict.get('query')}")
    print(f"⏱️ Response time: {response_dict.get('response_time')}s")
    print(f"\n📊 RESULTS ({len(response_dict.get('results', []))}):")
    print("-" * 50)
    
    for i, r in enumerate(response_dict.get('results', []), 1):
        print(f"\n{i}. {r.get('title')}")
        print(f"   URL: {r.get('url')}")
        print(f"   Content: {r.get('content')}")  # 👈 FULL content, no truncation
        print(f"   Score: {r.get('score')}")
        if r.get('raw_content'):
            print(f"   Raw: {r.get('raw_content')}")
    print("-" * 50)

# Use it
raw = web_search.invoke("best pasta recipes")
show_full_response(raw)

🔍 Query: best pasta recipes
⏱️ Response time: 0.77s

📊 RESULTS (5):
--------------------------------------------------

1. Our 32 Most Popular Pasta Recipes for Anytime Comfort Cravings
   URL: https://food52.com/story/23661-best-pasta-recipes-easy-dinner-ideas
   Content: Our 32 Best Pasta Recipes Ever. 1. Martha Stewart's One-Pan Pasta ; Victoria Granof's Pasta con Ceci · 4. Spicy Eggplant Pasta ; Ina Garten's Pasta
   Score: 0.89030397

2. Our Best Pasta Recipes - NYT Cooking
   URL: https://cooking.nytimes.com/topics/our-best-pasta-recipes
   Content: # Our Best Pasta Recipes. ### Easy Pasta Recipes. Smoky Tomato Pasta With Shrimp and Almonds. ### Smoky Tomato Pasta With Shrimp and Almonds. ### Vegan Pesto Pasta Salad. Pasta Salad With Sun-Dried Tomato Pesto. ### Pasta Salad With Sun-Dried Tomato Pesto. Pesto alla Trapanese (Pasta With Tomatoes, Almonds and Mint). ### Pesto alla Trapanese (Pasta With Tomatoes, Almonds and Mint). ### Pasta Sauce Recipes. ### Pasta Salad Recipes. ###

In [10]:
show_full_response(response)

🔍 Query: What are te best recepies in the world
⏱️ Response time: 0.86s

📊 RESULTS (5):
--------------------------------------------------

1. 22 'World's Best' Recipes That Our Community Can't Get Enough Of
   URL: https://www.allrecipes.com/worlds-best-recipes-8787129
   Content: Our list of 22 recipes includes everything from pork chops and soups to pastries, funnel cakes, and fudge, all claiming to be the best.
   Score: 0.6100127

2. Top 100 Traditional Recipes in the World - TasteAtlas
   URL: https://www.tasteatlas.com/best-recipes-in-the-world
   Content: Top 100 Traditional Recipes in the World · Vori-vori · Conchitas a la parmesana · Zhēngjiǎo · Karantika · Otoro nigiri sushi · Chutoro nigiri sushi.
   Score: 0.6002124

3. Best Recipes From Around the World - Forever Lost In Travel
   URL: https://foreverlostintravel.com/best-recipes-from-around-the-world/
   Content: That’s why I thought what better way than to collaborate with some amazing and worldly travel bloggers to bri

## PHASE 3: Agent Creation

In [11]:
sys = """

You are a personal chef. The user will give you a list of ingredients they have left over in their house.

Using the web search tool, search the web for recipes that can be made with the ingredients they have.

Return recipe suggestions and eventually the recipe instructions to the user, if requested.

"""

In [12]:
from langchain.agents import create_agent

agent = create_agent(
    model = model,
    tools = [web_search],
    system_prompt = sys
)



In [13]:
#Test the agent 

from langchain.messages import HumanMessage
response = agent.invoke(
{"messages": [HumanMessage(content="I have potatos, eggs, meat, cheese and some spices, what can i make")]})

pprint(response['messages'][-1].content)

[{'extras': {'signature': 'CuULAb4+9vv+yWLjTHxssOOTpNhM/rJTtdC6EyNXXfXW4dJdMjGP5GedMQB9SDfHqodCl9SPzUh51827bO8ctWpiM2hIsqSFLAfgngVZP1kDm2zXEnHCyWR9hoyjd+YrVU9Alaz4GKyLL+6cnUdq2OGqETPjVWxln2bCp8fIWYp38ZkUCJZSbrerI7bqnoXLfP8O84MR30mKVOEvTkHezJbgEF4OSmk1OzepERsgOwTN0QLdVX3CxOCBStmpb6FqMwnCoAZCETZHIQERLUiS3YmjnhpgrFys5D9u5gjzlYX0hEZ3SsxU4sH1hW5rQeB1DSz8cuqRP4vS04JSvskUlq5oFKeEPJiTEumKoLKKwpJ2v4TkGuRnEn6IEATIk23r/RSCylrWR6hPiDXcrVku3pSncPF3lkcGAnYX9cCoLjYTGH3qvqNr7V/q4NsX2In3UfnQiKq2V1r/LNZUryQijVQzsz9Ju0NB64UwUHgSZucorkSgeJR2Gr7mCYFulytLY+N7tkMd7iuQQJFKjv/cQ7wlLZutyoGHqDv2iluhAECMDWUs9g2GGJqCSvmECb1XqGuC9qn41XQ2Zawy5XttHYhPzcbKYl9Sqq+niLENSDzmW5H5TkdL4+qB0JZ8V8PJPHYbGWK9qJVmYlvvrUtA0ZfK8kd/gpw3ODRYNuZ4phsQ5Q5FuKVfJ4ZlGnApFu/zVTul6unP6GvZu8zn471wnTQGA7d+Nubz0jNgJNgQFSoiGjgnZXJo47BDFJjh3hlKOwRIH7qAEZPgbOunhYcWC+dAH9Gl27GQ+w0YOPADcC45sqGgupsm7pChyrw/Rd9LfB6hMfMf5AV98ZT2uto/kK9uE3T2JRGRpJuIfxEqv4xhuLmtXY84KvN+BaWVxHmwp3fz9u3saqRFn1+3udNAe3TisMSVGim4gkhGFiAzxoIM0CzMfbBNKc4mYCrMU95xJYNQu/CDC5TKV

In [14]:
pprint(response)

{'messages': [HumanMessage(content='I have potatos, eggs, meat, cheese and some spices, what can i make', additional_kwargs={}, response_metadata={}, id='97699f06-6302-4212-bd5a-336475650f5e'),
              AIMessage(content='', additional_kwargs={'function_call': {'name': 'web_search', 'arguments': '{"query": "recipes with potatoes, eggs, meat, cheese, spices"}'}, '__gemini_function_call_thought_signatures__': {'83272b94-23ae-4d48-a975-fd919c3950a6': 'Cq4CAb4+9vu3xKdwEKhWJVfnL65xpmzKoDuqk3lRe+Nr5PUJMxb48RUzrSmHTMmraj3laL0t1aoEt4PxWiLbmSlqJH7Mz78HgXLyhQPGHWtmFutK5fe+EpU9OFGmBb30jnB4+8h76mjgUQuzyE2Hck7RFLJaXP9aSxgV8Wc0bM8bdG6hXMsE/36oGIzFu5s8GPqRzN9vAdM6NLHfwjNboiooi8IDZUQEfi5Kg2ppvaR5iCJpXHSj/8iL0XIAMFBiULnM9qKZ4ASt4immTIvara4Fj4mHal5JGhogfm1HtjHJGdYbS5XOUjyWk5faFVs60AQJK4+svQifGqS6cY90syqg8/IIJiNynhGstYpl7Ua7pvINl1j9qb0Nons9mWMBa1zC8dUslDtC4zGx8lSuxc0='}}, response_metadata={'finish_reason': 'STOP', 'model_name': 'gemini-2.5-flash', 'safety_ratings': [], 'model_provider': 'google_gen

In [15]:
def get_clean_response(response):
    """Extract just the text response for your users"""
    messages = response['messages']
    
    # Get the last message (the final answer)
    last_message = messages[-1]
    
    # Handle different response formats
    if isinstance(last_message.content, list):
        # Gemini format: content is a list of dicts
        for item in last_message.content:
            if item.get('type') == 'text':
                return item['text']
    elif isinstance(last_message.content, str):
        # Simple string response
        return last_message.content
    
    return "Sorry, I couldn't understand the response"

# Use it:
clean_response = get_clean_response(response)
print(clean_response)

Based on the ingredients you have, here are a few recipe suggestions:

1.  **Hearty Southwestern Frittata with Potato and Beef:** This sounds like a great option as it uses potato, beef (your meat), eggs, cheese, and spices.
2.  **Cheesy Potato Egg Scramble:** This recipe typically includes potatoes, eggs, cheese, and a type of meat like ham (which could be substituted with your meat). It also uses spices.
3.  **One-Pot Spicy Eggs and Potatoes:** While this one focuses more on eggs and potatoes, it does include cheese and can be easily adapted to include your meat.

Would you like to hear more about any of these, or perhaps get the instructions for one of them?


In [16]:
# Check the second message (index 1) for tool_calls
if hasattr(response['messages'][1], 'tool_calls') and response['messages'][1].tool_calls:
    print("✅ Agent used a tool!")
    print(f"Tool used: {response['messages'][1].tool_calls[0]['name']}")
    print(f"Query: {response['messages'][1].tool_calls[0]['args']['query']}")
else:
    print("❌ Agent did NOT use a tool")

✅ Agent used a tool!
Tool used: web_search
Query: recipes with potatoes, eggs, meat, cheese, spices


In [17]:
response2 = agent.invoke({
    "messages": [HumanMessage(content="Yes, give me the instructions for the One-Pot Spicy Eggs and Potatoes recipe please")]
})
clean_response = get_clean_response(response2)
print(clean_response)

It looks like the "One-Pot Spicy Eggs and Potatoes" recipe is from Pinch of Yum!

To give you the instructions, I need a little more information. Could you please visit the link below and let me know if you'd like me to summarize the instructions for you?

[One-Pot Spicy Eggs and Potatoes Recipe - Pinch of Yum](https://pinchofyum.com/one-pot-spicy-eggs-and-potatoes)


In [18]:
response2 = agent.invoke({
    "messages": [HumanMessage(content="Yes, give me the instructions for first recipe now please")]
})
clean_response = get_clean_response(response2)
print(clean_response)

I would love to help you with that! However, you haven't told me what ingredients you have yet, and I haven't suggested any recipes. Please tell me what ingredients you have, and I'll find some recipes for you!


## ==> Agents won't hold a converstation unless we use checkpointers.

In [20]:
# === PHASE 3 ENHANCEMENT: Image Input for Fridge/Ingredients ===
# 1. Upload image 
from ipywidgets import FileUpload
from IPython.display import display
import base64

uploader = FileUpload(accept='.png,.jpg,.jpeg', multiple=False)
display(uploader)

FileUpload(value=(), accept='.png,.jpg,.jpeg', description='Upload')

In [22]:
# 2. Convert uploaded image to base64
def image_to_base64(uploader):
    uploaded_file = uploader.value[0]
    img_bytes = bytes(uploaded_file["content"])
    img_b64 = base64.b64encode(img_bytes).decode("utf-8")
    mime_type = uploaded_file["type"] # e.g. "image/jpeg"
    return img_b64, mime_type

# 3. Build the multimodal message (Gemini-compatible format)
def ask_agent_with_image(agent, uploader, text_prompt, config= None #we updated the function to accept  
                           # and pass the config 
                        ):
    img_b64, mime_type = image_to_base64(uploader)

    # Gemini needs this exact format for base64 images:
    question = HumanMessage(content=[
    {"type": "text", "text": text_prompt},
    {
        "type": "image_url",
        "image_url": {"url": f"data:{mime_type};base64,{img_b64}"}
    }
    ])

    response = agent.invoke({"messages": [question]}, config # we updated the function to accept and 
                           #pass the config
                           )
    return get_clean_response(response)

result = ask_agent_with_image(
agent,
uploader,
"Here's a photo of my fridge. What can I cook with these ingredients?"
)
print(result)


Based on the ingredients you have: cucumber, cherry tomatoes, spinach, red chili, leeks, potatoes, bell peppers, eggs, bread, milk, and lemon, here are some recipe suggestions:

**Main Dishes:**
*   **Potato, Spinach, Bell Pepper, and Leek Frittata:** This is a great way to use your eggs, potatoes, spinach, bell peppers, and leeks. You can add a kick with the red chili.
*   **Egg & Veggie Salad Bowl:** Combine hard-boiled eggs with cucumber, cherry tomatoes, spinach, and bell peppers for a refreshing salad. You can dress it with a lemon vinaigrette.
*   **Savory Bread Pudding/Strata:** Use your bread, eggs, milk, leeks, spinach, and bell peppers to create a savory bread pudding or strata. This is a comforting and versatile dish.

**Salads/Sides:**
*   **Tuscan Tomato, Bread, and Cucumber Salad (Panzanella-style):** A fresh salad using your bread, tomatoes, and cucumbers. You can add some bell peppers and a lemon dressing.
*   **Chopped Salad with Spinach, Cherry Tomatoes, Cucumber, and

## PHASE 4: Memory & Conversation

##### The problem is always true for LangChain agents — without a checkpointer, every .invoke() call starts with a blank memory.

In [23]:
from langchain.agents import create_agent
from langgraph.checkpoint.memory import InMemorySaver

agent = create_agent(
    model = model,
    tools = [web_search],
    system_prompt = sys,
    checkpointer=InMemorySaver()
    #The problem is always true for LangChain agents — 
    # without a checkpointer, every .invoke() call starts with a blank memory.
)

config = {"configurable": {"thread_id": "chef_session_1"}}

In [24]:
# Turn 1 — send image
result = ask_agent_with_image(
    agent,
    uploader,
    "Here's a photo of my fridge. What can I cook with these ingredients?",
    config=config  
)
print(result)

Based on the ingredients you have: potatoes, bell peppers, eggs, tomatoes, leeks, and bread, here are a few recipe ideas:

*   **Potato & Leek Frittata:** This sounds delicious and uses many of your ingredients. You can also add the bell peppers and tomatoes to it.
*   **Tomato, Leek, and Egg Casserole:** Another great option that combines these flavors.
*   **Egg, Potato and Tomato Breakfast/Dinner:** This is a versatile dish you can serve with your bread.

Would you like the instructions for any of these, or perhaps some other suggestions?


In [25]:
# Turn 2 — agent remembers the image and previous suggestions!
from langchain.messages import HumanMessage

response = agent.invoke(
    {"messages": [HumanMessage(content="Give me the full recipe for the frittata please")]},
    config 
)
print(get_clean_response(response))

Here's a full recipe for a Potato, Leek, Bell Pepper, and Tomato Frittata, adapting from a "Fresh Herb, Potato, and Goat Cheese Frittata" recipe to include your ingredients:

**Potato, Leek, Bell Pepper, and Tomato Frittata**

This frittata is a delicious and versatile dish that's perfect for breakfast, brunch, lunch, or a light dinner.

**Yields:** 4-6 servings
**Prep time:** 15 minutes
**Cook time:** 25-30 minutes

**Ingredients:**

*   **Potatoes:** 2-3 medium yellow-fleshed potatoes (like Yukon Gold), thinly sliced (about 1/8-inch thick)
*   **Leeks:** 1 large leek, white and light green parts only, thinly sliced and thoroughly washed
*   **Bell Pepper:** 1 medium bell pepper (any color, such as red or yellow), cored, seeded, and thinly sliced
*   **Tomatoes:** 1/2 cup cherry tomatoes, halved, or 1-2 small regular tomatoes, diced
*   **Eggs:** 8-10 large eggs
*   **Milk (optional):** 2 tablespoons milk or cream (for a creamier texture)
*   **Garlic:** 2 cloves garlic, minced (or 1 

## PHASE 5: ENHANCEMENTS 

A weak point observed in last phase is the Turn 2 response, instead of giving full instructions, it just returned a link. That's because the web search tool only grabbed a snippet(short preview), not the full page content.                  
The fix is to fetch the actual page content after finding the URL 

In [26]:
# Add this alongside your web_search tool
import requests
from bs4 import BeautifulSoup

@tool
def get_page_content(url: str) -> str:
    """Fetch the full content of a webpage, useful for getting complete recipe instructions"""
    try:
        headers = {"User-Agent": "Mozilla/5.0"}
        response = requests.get(url, headers=headers, timeout=10)
        soup = BeautifulSoup(response.text, "html.parser")
        
        # Remove noise (scripts, styles, nav bars etc)
        for tag in soup(["script", "style", "nav", "footer", "header"]):
            tag.decompose()
        
        # Get clean text
        text = soup.get_text(separator="\n", strip=True)
        
        # Limit to avoid overwhelming the model
        return text[:5000]
    
    except Exception as e:
        return f"Could not fetch page: {str(e)}"

The User-Agent header is a workaround — some websites block requests that don't look like a real browser, so we fake it. But yes, it varies page to page — some sites will still block you (like big ones with anti-scraping protection e.g. NYT Cooking), some won't. It's not 100% reliable but works for most recipe sites.

In [27]:
# Update the system prompt

sys = """
You are a personal chef. The user will give you a list of ingredients they have,
either as text or as a photo of their fridge.

You have two tools:
- web_search: use this to find recipe suggestions based on ingredients
- get_page_content: use this to fetch the FULL instructions of a specific recipe URL

When the user asks for full recipe instructions, ALWAYS use get_page_content 
on the recipe URL to retrieve the complete steps — never just return a link.
"""



In [28]:
# Update the Agent 
from langchain.agents import create_agent
from langgraph.checkpoint.memory import InMemorySaver

agent = create_agent(
    model=model,
    tools=[web_search, get_page_content],  # We add the new tool
    system_prompt=sys,
    checkpointer=InMemorySaver()
)

### Redo the same tests :


In [29]:
# Reset config (new agent = fresh memory)
config = {"configurable": {"thread_id": "chef_session_2"}}

In [30]:
# Turn 1 — image
result = ask_agent_with_image(
    agent,
    uploader,
    "Here's a photo of my fridge. What can I cook with these ingredients?",
    config=config
)
print(result)

Based on the ingredients in your fridge, which include bread, eggs, potatoes, bell peppers (red and yellow), tomatoes, leeks, cucumber, orange, lemon, red chili peppers, and some leafy greens, here are a few recipe ideas:

1.  **Scrambled Eggs with Peppers, Tomatoes, and Potatoes:** This is a versatile and hearty option. You can sauté the bell peppers, tomatoes, and diced potatoes, then add the eggs and scramble them all together. You can also incorporate the leeks and a bit of the chili for extra flavor. Serve with the bread.
2.  **Egg and Leek Bake/Frittata:** You could make an egg bake or a frittata using the eggs, sautéed leeks, bell peppers, and tomatoes. Potatoes could be roasted and added in too.
3.  **Hearty Breakfast Skillet:** A one-pan dish with diced potatoes, bell peppers, tomatoes, and leeks, topped with fried or scrambled eggs and served with toast.

Would you like the full recipe for "Scrambled Eggs With Peppers, Tomatoes and Potatoes" from NYT Cooking?


In [31]:
# Turn 2 — should now return FULL instructions, not just a link
response = agent.invoke(
    {"messages": [HumanMessage(content="Give me the full recipe for the frittata please")]},
    config
)
print(get_clean_response(response))

I found a "Potato and Leek Frittata" recipe from Inspiralized that seems promising, as it uses several of your ingredients. There's also a general "Easy Frittata Recipe & Variations" from Sally's Baking Addiction that might be adaptable.

Would you like the "Potato and Leek Frittata" recipe from Inspiralized or the "Easy Frittata Recipe & Variations" from Sally's Baking Addiction? Or would you prefer I search for a recipe that explicitly includes bell peppers and tomatoes in addition to the leeks and potatoes?


In [32]:
response = agent.invoke(
    {"messages": [HumanMessage(content="Yes, give me the Potato and Leek Frittata from Inspiralized")]},
    config
)
print(get_clean_response(response))

ChatGoogleGenerativeAIError: Error calling model 'gemini-2.5-flash' (RESOURCE_EXHAUSTED): 429 RESOURCE_EXHAUSTED. {'error': {'code': 429, 'message': 'You exceeded your current quota, please check your plan and billing details. For more information on this error, head to: https://ai.google.dev/gemini-api/docs/rate-limits. To monitor your current usage, head to: https://ai.dev/rate-limit. \n* Quota exceeded for metric: generativelanguage.googleapis.com/generate_content_free_tier_requests, limit: 20, model: gemini-2.5-flash\nPlease retry in 58.138054354s.', 'status': 'RESOURCE_EXHAUSTED', 'details': [{'@type': 'type.googleapis.com/google.rpc.Help', 'links': [{'description': 'Learn more about Gemini API quotas', 'url': 'https://ai.google.dev/gemini-api/docs/rate-limits'}]}, {'@type': 'type.googleapis.com/google.rpc.QuotaFailure', 'violations': [{'quotaMetric': 'generativelanguage.googleapis.com/generate_content_free_tier_requests', 'quotaId': 'GenerateRequestsPerDayPerProjectPerModel-FreeTier', 'quotaDimensions': {'location': 'global', 'model': 'gemini-2.5-flash'}, 'quotaValue': '20'}]}, {'@type': 'type.googleapis.com/google.rpc.RetryInfo', 'retryDelay': '58s'}]}}

## ⚠️ Important: Gemini Free Tier Limit

This project uses **Google Gemini 2.5 Flash** which has a free tier limit of **20 requests per day**.

Each agent call counts as one or more requests depending on how many tool calls are made.  
If you hit the limit you will see a `429 RESOURCE_EXHAUSTED` error — just wait until the next day and the quota resets automatically.

**To avoid hitting the limit quickly:**
- Don't re-run cells unnecessarily
- Test with text input first before using image input
- Each image turn costs more requests than text turns